### PR TITLE
feat: Fastify support

### PR DESCRIPTION
**Pull request by Mr0Bread**
- [x] Add docs

### CODE REVIEW

File: packages/uploadthing/src/internal/types.ts [hunk 1]
```diff
@@ -25,8 +26,11 @@ export type MaybePromise<TType> = TType | Promise<TType>;
...
 export type Overwrite<T, U> = Omit<T, keyof U> & U;
 
+// Omitting body because it's not used and is represented
+// by different types in different implementations
+// Anyway, we use json() instead to get the body of the request
 export type RequestLike = Overwrite<
-  WithRequired<Partial<Request>, "json">,
+  Omit<WithRequired<Partial<Request>, "json">, "body">,
   {
     headers: Headers | IncomingHttpHeaders;
```
Comment (juliusmarminge) [inline packages/uploadthing/src/internal/types.ts:35]: i think it's a bit better ts perf to put this inside the overwrite object (skips 1 Omit)

```ts
export type RequestLike = Overwrite<
  WithRequired<Partial<Request>, "json">,
  {
    body?: never; // we only use `.json`, don't need `body`
    headers: Headers | IncomingHttpHeaders;
  }
>;

**Review by t3dotgg — APPROVED** [review]
This PR was a great reminder that I love fastify

File: packages/uploadthing/src/internal/fastify.ts [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+import type { FileRouter, inferErrorShape } from "./types";
+
+export const fastifyUploadthingPlugin = <TRouter extends FileRouter>(
+  fastify: FastifyInstance,
+  opts: RouterWithConfig<TRouter>,
+  done: (err?: Error) => void,
+) => {
+  const requestHandler = buildRequestHandler<TRouter, "fastify">(opts);
+
+  fastify.register;
```
Comment (Mr0Bread) [inline packages/uploadthing/src/internal/fastify.ts:18]: Looks like some kind of leftover, have to be removed

File: packages/uploadthing/src/internal/fastify.ts [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+
+    if (response.status !== 200) {
+      // We messed up - this should never happen
+      void res
+        .status(500)
+        .headers({
+          "x-uploadthing-version": UPLOADTHING_VERSION,
+        })
+        .send("An unknown error occured");
+    }
```
[the remark this answers is not in the document]
Comment (Mr0Bread) [reply packages/uploadthing/src/internal/fastify.ts:57]: Yep, you are totally right, will fix. Thank you

**Review by juliusmarminge — APPROVED** [review]
Looks solid. Why does the express example have a frontend and this one doesn't?

**Comment by Mr0Bread** [comment]
> Looks solid. Why does the express example have a frontend and this one doesn't?

Will add it

**Review by juliusmarminge — APPROVED** [review]

**Comment by juliusmarminge** [comment]
Wtf why isn't the backend adapter section in the sidebar?

https://docs-uploadthing-qt60mspp8-pinglabs.vercel.app/backend-adapters/fastify

**Comment by Mr0Bread** [comment]
> Wtf why isn't the backend adapter section in the sidebar?
> 
> https://docs-uploadthing-qt60mspp8-pinglabs.vercel.app/backend-adapters/fastify

I see that locally everything is fine. I guess preview deployment is funny. I already had cases when it didn't display proper state of website even though the code is fine

**Comment by juliusmarminge** [comment]
But prod should show express then: https://docs.uploadthing.com/backend-adapters/express

**Comment by Mr0Bread** [comment]
Yeah, you are right. Something is off. Maybe deployment on Vercel didn't occur or it got interrupted? Or outdated version is being served?

**Comment by markflorkowski** [comment]
> Yeah, you are right. Something is off. Maybe deployment on Vercel didn't occur or it got interrupted? Or outdated version is being served?

I figured it out. It's a caching bug.

**Pull request merged**